src/ec/eddsa.rs

```diff
@@ -176,10 +176,56 @@ impl signature::VerificationAlgorithm for EdDSAParameters {
         }
         let msg = msg.as_slice_less_safe();
         let signature = signature.as_slice_less_safe();
-        bssl::map_result(unsafe {
-            GFp_ed25519_verify(msg.as_ptr(), msg.len(), signature.as_ptr(),
-                               public_key.as_ptr())
-        })
+
+        let mut a = Point::new();
+        if (signature[63] & 224) != 0 {
+            return Err(error::Unspecified);
+        }
+        let ret = unsafe {
+            GFp_x25519_ge_frombytes_vartime(&mut a, public_key.as_ptr())
+        };
+        if ret != 0 { return Err(error::Unspecified) }
+
+        for i in 0 .. ELEM_LIMBS {
+            a.x[i] = -a.x[i];
+            a.t[i] = -a.t[i];
+        }
+        let mut r_copy = [0u8; ELEM_LEN];
+        unsafe {
+            core::ptr::copy_nonoverlapping(signature.as_ptr(),
+                                           r_copy.as_mut_ptr(), ELEM_LEN)
+        };
```

Comment:
Again, I understand what you're trying to do regarding the direct transliteration and then improvement, but it's too much work to review the direct transliteration separately when it results in severely non-idiomatic Rust like this. In particular, we should use normal and safe array/slice copying here, instead of the unsafe `copy_nonoverlapping.